Code/Framework/AzToolsFramework/AzToolsFramework/UI/DPEDebugViewer/DPEDebugModel.cpp

```diff
@@ -0,0 +1,374 @@
+        DPEModelNode* assignedParent = (isColumn ? GetParentNode() : this);
+        newChild->m_parent = assignedParent;
+        newChild->m_displayString = value;
+
+        if (isColumn)
+        {
+            newChild->m_columnParent = this;
+            m_columnChildren.push_back(newChild);
+            const int numCols = m_columnChildren.size() + 1;
+            if (assignedParent->GetMaxChildColumns() < numCols)
+            {
+                assignedParent->m_maxChildColumns = numCols;
+            }
+        }
+        else
+        {
+            m_children.push_back(newChild);
+        }
+
+        return newChild;
+    }
+
+    AZ::Dom::Value DPEModelNode::GetValue() const
+    {
+        QVector<size_t> reversePath;
+        const DPEModelNode* currNode = this;
+
+        // traverse upwards to the root determine the path to the current node
+        bool traversingUpwards = true;
+        while (traversingUpwards)
+        {
+            if (currNode->m_columnParent)
+            {
+                reversePath << currNode->m_domValueIndex;
+                currNode = currNode->m_columnParent;
+            }
+            else if (currNode->m_parent)
+            {
+                reversePath << currNode->m_domValueIndex;
+                currNode = currNode->m_parent;
+            }
+            else
+            {
+                traversingUpwards = false;
+            }
+        }
+        // start at the top of the adapter and move down the inverse of the path we took traversing upwards
+        auto returnValue = GetModel()->GetAdapter()->GetContents();
+        for (auto reverseIter = reversePath.rbegin(); reverseIter != reversePath.rend(); ++reverseIter)
+        {
+            returnValue = returnValue[*reverseIter];
+        }
+
+        return returnValue;
+    }
+
+    DPEDebugModel::DPEDebugModel(QObject* parent)
+        : QAbstractItemModel(parent)
+    {
+    }
+
+    void DPEDebugModel::SetAdapter(AZ::DocumentPropertyEditor::DocumentAdapter* theAdapter)
+    {
+        m_adapter = theAdapter;
+        m_isResetting = (m_rootNode != nullptr);
+        if (m_isResetting)
+        {
+            beginResetModel();
+            m_rootNode->deleteLater(); // recursively destroys the existing tree
+        }
+
+        // recursively populate the node tree with the new adapter information
+        auto theContents = theAdapter->GetContents();
+        m_rootNode = new DPEModelNode(DPEModelNode::NodeType::RootNode, 0, this);
+        m_rootNode->Populate(theAdapter->GetContents());
+
+        if (m_isResetting)
+        {
+            endResetModel();
+            m_isResetting = false;
+        }
+        else
+        {
+            // inform any listening views that top-level rows have been added,
+            // they will automatically traverse the child nodes
+            const int numRootChildren = m_rootNode->GetChildCount();
```

Comment:
What does avoiding the model reset here get us? I've generally always done a reset model on populate, figuring the view has to query the whole structure anyways, but I've never actually profiled this, just seems like maybe we could get rid of `m_isResetting` entirely and streamline this.